### ISSUE TITLE
Simple GroupMonitor example?

### ISSUE THREAD

**Issue by StefanOltmann** [pitschr/knx-link#82]
Can you provide an example for usage of the GroupMonitor that's as simple as this sample code from calimero-project?

https://github.com/calimero-project/introduction/blob/master/src/main/java/GroupMonitor.java

Maybe you should add this also to your docs.

I find the tty monitor demo a bit complex.

**Comment by pitschr** [comment]
Hi Stefan.

Given your example, it sounds very similiar with the "plugin" example:
https://github.com/pitschr/knx-link/blob/master/knx-examples/src/main/java/li/pitschmann/knx/examples/plugin/Main.java
and
https://github.com/pitschr/knx-link/blob/master/knx-examples/src/main/java/li/pitschmann/knx/examples/plugin/MyPlugin.java

It is not same like Calimero because knx-link is designed to work with plugin(s). To be more similiar with the example from Calimero you may try out:

```
public class MyPlugin implements ObserverPlugin {

    @Override
    public void onInitialization(final KnxClient client) {
        System.out.println("Initialized by client: " + client);
    }

    @Override
    public void onIncomingBody(final Body item) {
        if (item instanceof TunnelingRequestBody) {
            final var cemi = ((TunnelingRequestBody)item).getCEMI();
            System.out.println(LocalTime.now() + " " + cemi.getSourceAddress().getAddress() + "->" + ((GroupAddress)cemi.getDestinationAddress()).getAddressLevel3() + " read.res/write.ind(" + cemi.getApci().getFriendlyName() + "): " + item.getRawDataAsHexString());
        }
        // ignore other service types (e.g. search, description, connect, disconnect)
    }

    @Override
    public void onOutgoingBody(final Body item) {
        if (item instanceof TunnelingRequestBody) {
            final var cemi = ((TunnelingRequestBody)item).getCEMI();
            System.out.println(LocalTime.now() + " " + cemi.getSourceAddress().getAddress() + "->" + ((GroupAddress)cemi.getDestinationAddress()).getAddressLevel3() + " read.req(" + cemi.getApci().getFriendlyName() + "): " + item.getRawDataAsHexString());
        }
        // ignore other service types (e.g. search, description, connect, disconnect)
    }

    @Override
    public void onError(final Throwable throwable) {
        System.out.println("On Error: " + throwable.getMessage());
    }
}
```

And the main class (just for completness):
```
public class Main {

    public static void main(final String[] args) {

        final var config = ConfigBuilder
                .tunneling()                   // communication mode: tunneling
                .plugin(MyPlugin.class)        // register MyPlugin
                .build();                      // create immutable config

        // create KNX client and connect to KNX Net/IP device using auto-discovery
        try (final var client = DefaultKnxClient.createStarted(config)) {
            final var groupAddress = GroupAddress.of(1,1,113);
            client.readRequest(groupAddress);
            System.out.println("Status: " + client.getStatusPool().getStatusFor(groupAddress));

            // end monitor after 10 seconds
            Sleeper.seconds(10);

            // then
            System.out.println("Time is up!");
        }

        // auto-closed and disconnected by KNX client
    }

}
```

This will create following output (in my case fetching the status from KNX Group Address `1/2/113`):
```
Initialized by client: li.pitschmann.knx.core.communication.DefaultKnxClient@35461a77
20:55:03.193087200 0.0.0->1/1/113 read.req(Group Value Read): 0x04 37 00 00 11 00 BC E0 00 00 09 71 01 00 00
20:55:03.202852900 15.15.242->1/1/113 read.res/write.ind(Group Value Read): 0x04 37 00 00 2E 00 BC E0 FF F2 09 71 01 00 00
20:55:03.248849 1.0.10->1/1/113 read.res/write.ind(Group Value Response): 0x04 37 01 00 29 00 BC E0 10 0A 09 71 01 00 40
Status: KnxStatusData{dirty=false, timestamp=2020-12-26T19:55:03.250850600Z, sourceAddress=IndividualAddress{addressType=AddressType{name=INDIVIDUAL, friendlyName=Individual Address, code=0 (0x00)}, address=1.0.10, rawData=0x10 0A}, apci=APCI{name=GROUP_VALUE_RESPONSE, friendlyName=Group Value Response, code=64..127 (0x40..0x7F)}, apciData=[0] (0x00)}
... lines omitted ...
Time is up!
```

Hope that is helpful for you.

You may also look at the [File Audit Plugin](https://github.com/pitschr/knx-link/tree/master/knx-core-plugins/audit) which prints out all packets to the file (default output format: JSON)

The TTY Monitor Demo project is mainly bit complex because it operates with ANSI Escape Characters which allows to print out the KNX traffic in a nice way. Not supported by Windows machines AFAIK.

If you want to look for other examples, check out this folder:
https://github.com/pitschr/knx-link/tree/master/knx-examples/src/main/java/li/pitschmann/knx/examples

Let me know if something is unclear. KR, Christoph

**Comment by StefanOltmann** [comment]
Hi Christoph,

that helps a lot. Thank you.

The monitoring also works if I leave out the initial read request as my weather station produces a lot of KNX traffic all the time. :)

Now I try to figure out how I get the status value from the item.
I don't really have a deep understanding of KNX internals, so "cEMI", "APCI" and so on don't ring a bell. :/

From the documentation I understand that I have to call "DPT1.SWITCH.of()" to get the value of my switches.
But how can I get to the byte array that I need to pass in there?

At my current stage I read switch values (booleans) and percentages (of dimmers, DPT8.PERCENT I guess).

**Comment by pitschr** [comment]
cEMI stands for common external message interface which is used as transport format on KNX Net/IP. This protocol is a part of KNX specification and it is flexible with optional fields.  APCI stands for Application Layer Protocol Control Information and contains the data to be transported. The KNX Client should take care of both and should not be in your domain.

To get a value you have to send a "READ" request to a KNX Group Address (that contains the READ flag, defined in your ETS project) first:

```
var groupAddress = GroupAddress.of(1,2,113);
client.readRequest(groupAddress); 
```

As soon the read request has been sent to the KNX Net/IP device (=router or interface), you will get an acknowledge message (will be handled by the KNX client too). The KNX Net/IP device will forward the request to the corresponding KNX component that listens on the given KNX Group Address. As soon the KNX component got the read request flag, and the group address has the read flag set as well (=allows to respond you), it will send the response to the KNX Net/IP device and the KNX Net/IP device will forward it to you.

To get the value use
```
var knxStatusPool = client.getStatusPool();
knxStatusPool.getValue(groupAddress, DPT1.SWITCH).getBooleanValue(); // returns bool value
```

To send the value to a KNX Group Address use:
```
var boolValue = DPT1.SWITCH.of(true);
client.writeRequest(groupAddress, boolValue);
```

You don't need to translate it to byte-array (and back) at all. KNX Client will do the hard job for you. In case you have to deal with a DataPointType (DPT) which is not implemented by the KNX client, you have the ability to send data by byte-array. This is only the exception where it would make sense to deal with byte-array directly. For the sake of clarification, both have same effect:
```
// better - as it is implemented and tested based on KNX specification
client.writeRequest(groupAddress, DPT1.SWITCH.of(true));

// try to avoid
client.writeRequest(groupAddress, new byte[] { 0x01 });
```

Same applies to the other Data Point Types like DPT8 etc. In case you don't know which KNX Group Address has READ flag enabled, check out your project in ETS. If the KNX Group Address has no READ flag set, you won't get a response. If you are using a KNX Filter / Line Coupler, make sure that the traffic for Group Address is also permitted. If you are not sure which DataPointType to be used, also take a look in your ETS project.

I suggest to take a look at this page:
https://github.com/pitschr/knx-link/tree/master/knx-core
(The section "Example: Inverse the lamp status" covers both: READ and WRITE)

**Comment by StefanOltmann** [comment]
Thank you for your detailed exlanation!

So this means I cannot extract the status value from "Body item" and instead I need to get it from knxStatusPool.

With calimero-project the new value is also part of the GroupMonitor-event-parameter-object. It's just really ugly to parse it.

Is the call to knxStatusPool efficent? Is this a cache? Or will it cause new traffic on the bus?

My use case is to cache the last known status of each device, because a GUI gets this to show the current state of everything without a expensive read request to each device everytime the GUI is opened.

**Comment by pitschr** [comment]
My examples from yesterday were the ways how to get the data outside of plugin. Of course, you can extract the status value from `Body item` (within plugin itself). But to extract the status properly, you need to know which Data Point Type the data belongs to. It is not a part of KNX packet as KNX is designed to send/receive packets with most minimum only. That is the reason why it is also efficient with only 9600 bit/s. If you register the KNXPROJ file (that one used by the ETS) to the KNX Client you will also get KNX Group Address information and then it can decode it for you automatically using `DataPointRegistry` class.

Here an example how to extract the value by yourself:
```
    public void onIncomingBody(final Body item) {
        if (item instanceof TunnelingRequestBody) {
            final var apciData = ((TunnelingRequestBody)item).getCEMI().getApciData();

            final var value = DPT8.PERCENT.of(apciData).getSignedValue(); // -327,68 .. 327,67 (according KNX DPT 8.010)
            final var unit = DPT8.PERCENT.getUnit(); // % (according KNX DPT 8.010)

            // ... do some stuff
        }
    }
```

Here an example how to extract the value and unit with help from KNXPROJ file (DPT1..DPT23 are supported):
````
    public void onIncomingBody(final Body item) {
        if (item instanceof TunnelingRequestBody) {
            final var cemi = ((TunnelingRequestBody)item).getCEMI();
            final var apciData = cemi.getApciData();
            final var groupAddress = (GroupAddress)cemi.getDestinationAddress();

            // read the data point type from KNX Project file
            final var dataPointTypeFromKNXPROJ = client.getConfig().getProject().getGroupAddress(groupAddress).getDataPointType();
            // auto-resolve to the right data point type
            final var dataPointType = DataPointRegistry.getDataPointType(dataPointTypeFromKNXPROJ);

            final var value = dataPointType.of(apciData).toText(); // get value as string (e.g. 12.34)
            final var unit = dataPointType.getUnit(); // get unit (e.g. % for percent)

            // ... do some stuff with "value" and "unit"
        }
    }
```

The `KnxStatusPool` is a wrapped hash map whereas the key is `KnxAddress` (either Individual Address or Group Address) and the value is `KnxStatusData` and contains all known status, either:
* it has been requested explicitly (e.g. using `client.readRequest(..)` )
* it has been auto-captured since the KNX client is started

The `KnxStatusPool` will not cause new traffic. If the status is not registered/known to the pool, you will get `null` otherwise you will get a `KnxStatusData`. This is because KNX packets are not responded immediately, they are acknowledged only. Using KNX you are only requesting for e.g. status, and as soon the response is arrived to the KNX client, the `KnxStatusPool` will be automatically updated with the last known status.

To achieve your use case you can re-use the `KnxStatusPool` to get the last known status. If you get `null` then you know that you have to send a read request first, and then you can try to get the last known status from `KnxStatusPool`. To make it more efficient you can use `existsStatusFor(KnxAddress)` before `getStatusFor(KnxAddress)` because it will respond it to you immediately if the last known status for the given `KnxAddress` exists at all. Based on that you know if you have to send a READ request or not. 

Keep in mind that it is not guaranteed that the KNX device will also respond to you (e.g. because of missing communication flag or read flag or wrong KNX Group Address was sent) and you should not be stuck in an infinite loop. Therefore, it is good idea to give up after e.g. 3 calls and throw an exception / or ignore it. When using `client.readRequest(GroupAddress)` then KNX client will do the hard job for you already. The returned boolean value is `true` if the acknowledge was successful. This method is a blocking-operation. To use a non-blocking operation use the `client.send(Body)` or `client.send(Body, timeout)` instead.

**Comment by StefanOltmann** [comment]
Again thank you for this very helpful information.

Manual extraction of the value providing the correct Data Point Type works.
All in all it's way cleaner and better than the way you do it with calimero-project. You did a great job on that!
The only problem for people like me with little insight into KNX internals is that you need to know what APCI is and that you need to get it there.

I suggest you provide some convenience methods to make this easier to find and use.
The Body object also could have a getValue() similar to the knxStatusPool for example.

I also tried the KNXPROJ version. I like very much how it automaticly selects the correct file without any configuration.
Unfortunately I can't use this way from Kotlin.

The line "DataPointRegistry.getDataPointType(dataPointTypeFromKNXPROJ);" gives an error:
"Not enough information to infer type variable T"

Is there a way without the generics to achieve the same thing?

**Comment by pitschr** [comment]
I'm not familiar with Calimero project. But thanks and it is good to know that the way is cleaner.

I can imagine to create a Helper / an Utility / `AbstractPlugin` class with common methods class to provide a better support for DataPointType rather than extending e.g. `TunnelingRequestBody` (and `RoutingIndicationBody` for routing) because both bodies reflects with the CEMI (and related APCI data) structure from KNX Specification paper. I may have to rethink tomorrow. Would that be suitable? What do you think? 

I am not familiar with the Kotlin and its error message "Not enough information to infer type variable T". But may you try out if it would solve the problem when changing the method signature for `li.pitschmann.knx.core.datapoint.DataPointRegistry#getDataPointType(java.lang.String)`

from
```
    public static <T extends DataPointType> T getDataPointType(final String id) {
```
to
```
    public static <T extends DataPointType<? extends DataPointValue<T>>> T getDataPointType(final String id) {
```

The reason is that `DataPointType` is a raw type. It is sufficient for Java. But not sure if it is the reason why Kotlin is struggling with it. Otherwise you may to need to cast it by yourself or can you advice how the method can be done to be compatible with Kotlin? Thanks.

**Comment by StefanOltmann** [comment]
Regarding the convience method for accessing the values I'm also not sure what would fit here. Not at the moment.

The suggested signature changes did not help.

If I make the Map public and access it directly the assigned value is of Kotlin type `DataPointType<out DataPointValue<*>>?`

It compiles in Kotlin if I take out the generic:
`    public static DataPointType getDataPointType(final String id) {`

Maybe you could provide a second method without generic for usage with Kotlin.

Unfortunately I don't know enough Kotlin to tell you if this is a bug or what the reason is here.
Currently I'm on version 1.4.10

**Comment by pitschr** [comment]
OK, I have some ideas how to get rid off the generics partially. Currently the infer type between the `DataPointType` and `DataPointValue` is bi-directional. I will make it uni-directional. But requires some refactoring. I guess it should be OK, because the library has no public version yet and that would improve the library further.

Will let you know as soon the preliminary version is available.

**Comment by pitschr** [comment]
I have created a separate branch called "master-2020-12-29". May you check if that would work for you?
https://github.com/pitschr/knx-link/tree/master-2020-12-29

Refactorings:
* No Generic type for DataPointType and DataPointValue anymore (they are now used on BaseDataPointType)
* for CEMI component: renamed `getApci()` to `getAPCI()`, `getTpci()` to getTPCI(), `getTpciPacketNumber()` to `getPacketNumber()`, `getApciData()` to `getData()`, `getNpduLength()` to `getLength()`

The signature is now (no wildcard anymore)
```
public static <T extends DataPointType> T getDataPointType(final String id)
```

Hope that works for Kotlin as the the infer type has been simplified much more.
Let me know if that works and if you need further input.

**Comment by StefanOltmann** [comment]
I very much like your refactoring.  Without the complicated generics structure and the better naming it's better understandable.

I can now write:
```
        override fun onIncomingBody(item: Body) {
            if (item is TunnelingRequestBody) {

                val groupAddress = item.cemi.destinationAddress as GroupAddress
                val dataPointTypeFromKNXPROJ = client!!.config.project.getGroupAddress(groupAddress)!!.dataPointType
                val dataPointType : DataPointType = DataPointRegistry.getDataPointType(dataPointTypeFromKNXPROJ)

                val value = dataPointType!!.of(item.cemi.data).toText() // get value as string (e.g. 12.34)
                val unit = dataPointType!!.unit // get unit (e.g. % for percent)

                println("$dataPointTypeFromKNXPROJ | $dataPointType | value = $value | unit = $unit")
            }
            // ignore other service types (e.g. search, description, connect, disconnect)
        }
```

Instead of `val dataPointType : DataPointType` I can write `val dataPointType : DPT5` or something else to specify T.
Before that IDEA suggested to write something cryptic as type, but it never compiled for me.

So this code is usable now! :)

It captures correctly the values of my dimmers and my NTP server:
```
DPST-5-1 | DPT5{id=5.001, description=Scaling (%)} | value = 40 | unit = %
DPST-11-1 | DPT11{id=11.001, description=Date} | value = 2020-12-30 | unit = 
DPST-10-1 | DPT10{id=10.001, description=Time Of Day} | value = Mittwoch, 20:24:00 | unit = 
```

I have trouble with my switches, but I will check my ETS file for possible errors first. :)

**Comment by StefanOltmann** [comment]
What would you suggest to run the monitoring infinite?

Your sample lets the monitor run for 10 seconds and I feel that I just don't want to hack that with Long.MAX_VALUE...

**Comment by pitschr** [comment]
That is great news.

If you really want to have monitoring infinite then I propose to take a look at this line:
https://github.com/pitschr/knx-demo-tty-monitor/blob/10337f6143f73b0816c3812b0ce224cf94955a39/src/main/java/li/pitschmann/knx/examples/tty/Main.java#L128

Here I am checking if the client is still running and also check if the time is over due. In your case you can only check if the client is still running. However, you need to implement how to stop the client. It can be done using a quit signal.

If you kill the java process (e.g. using CTRL+C, kill -9 command, etc.) then KNX client won't close the connection to your KNX Net/IP router properly. If that happens too frequently in short time, then it could happen that all connections slots are used until your KNX router/interface notices that the client is not responsive anymore and cleans up by themselves. Therefore, better close the KNX client properly to free the connection on your KNX router/interface so that the free connection slot can be re-used immediately.

Thanks for confirming it, I'll merge the changes of `master-2020-12-29` branch to the `master` branch and release it with version 0.0.5. Should happen this week as soon I cleaned up some test suites as well.

**Comment by pitschr** [comment]
merge done to master branch. I released it as `0.1.0` . The library has already been released to maven repository, however, it takes some time (few hours) until it is public visible.

I consider this task as closed. Please reopen or create a new issue in case something is still missing.
